rest-service/manager_rest/rest/resources_v3_1/operations.py

```diff
@@ -84,21 +87,76 @@ def put(self, operation_id, **kwargs):
     @authorize('operations')
     @marshal_with(models.Operation)
     def patch(self, operation_id, **kwargs):
-        request_dict = get_json_and_verify_params(
-            {'state': {'type': text_type}}
-        )
+        request_dict = get_json_and_verify_params({
+            'state': {'type': text_type},
+            'result': {'optional': True},
+            'exception': {'optional': True},
+            'exception_causes': {'optional': True},
+        })
         sm = get_storage_manager()
         with sm.transaction():
             instance = sm.get(models.Operation, operation_id, locking=True)
             old_state = instance.state
             instance.state = request_dict.get('state', instance.state)
+            self._insert_event(
+                instance,
+                request_dict.get('result'),
+                request_dict.get('exception'),
+                request_dict.get('exception_causes')
+            )
             if not instance.is_nop and \
-                    old_state not in TERMINATED_STATES and \
-                    instance.state in TERMINATED_STATES:
+                    old_state not in common_constants.TERMINATED_STATES and \
+                    instance.state in common_constants.TERMINATED_STATES:
                 self._modify_execution_operations_counts(instance, 1)
             instance = sm.update(instance, modified_attrs=('state',))
         return instance
 
+    def _insert_event(self, operation, result=None, exception=None,
+                      exception_causes=None):
+        if operation.type != 'RemoteWorkflowTask':
+            return
+        if not current_execution:
+            return
+        try:
+            context = operation.parameters['task_kwargs']['kwargs'][
+                '__cloudify_context']
+        except (KeyError, TypeError):
+            return
+        if exception is not None:
+            operation.parameters.setdefault('error', str(exception))
+        current_retries = context.get('current_retries') or 0
+        total_retries = context.get('total_retries') or 0
+
+        try:
+            message = common_events.format_event_message(
+                operation.name,
+                operation.state,
+                result,
+                exception,
+                current_retries,
+                total_retries,
+            )
+            event_type = common_events.get_event_type(operation.state)
+        except RuntimeError:
```

Comment:
What RuntimeErrors are we expecting that we don't want to log?

Reply:
missing (ignored) event type, thrown in https://github.com/cloudify-cosmo/cloudify-common/blob/78c8d7c2e6283135ce51b8d98ab6de95cf1bffae/cloudify/workflows/events.py#L70

Reply:
Hmm. Maybe we should've made that a more specific exception, but not a massive impact here at the moment.